### PR TITLE
GG-35191 Borrow numa allocator code from ignite

### PR DESCRIPTION
**Pull request by ademakov**

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/mem/unsafe/UnsafeMemoryAllocator.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
+/*
+ * Copyright 2019 GridGain Systems, Inc. and Contributors.
```
Comment (SammyVimes) [inline modules/core/src/main/java/org/apache/ignite/internal/mem/unsafe/UnsafeMemoryAllocator.java:2]:
```suggestion
 * Copyright 2022 GridGain Systems, Inc. and Contributors.
```

File: modules/core/src/main/java/org/apache/ignite/mem/MemoryAllocator.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright 2019 GridGain Systems, Inc. and Contributors.
```
Comment (SammyVimes) [inline modules/core/src/main/java/org/apache/ignite/mem/MemoryAllocator.java:2]:
```suggestion
 * Copyright 2022 GridGain Systems, Inc. and Contributors.
```

File: modules/numa-allocator/pom.xml [hunk 3]
```diff
@@ -0,0 +1,210 @@
+<?xml version="1.0" encoding="UTF-8"?>
+
+<!--
+ Copyright 2019 GridGain Systems, Inc. and Contributors.
```
Comment (SammyVimes) [inline modules/numa-allocator/pom.xml:4]:
```suggestion
 Copyright 2022 GridGain Systems, Inc. and Contributors.
```

File: modules/numa-allocator/pom.xml [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+                </os>
+            </activation>
+            <properties>
+                <cmake.classifier>linux-x86_64</cmake.classifier>
+            </properties>
+            <build>
+                <plugins>
+                    <plugin>
+                        <groupId>com.googlecode.cmake-maven-project</groupId>
+                        <artifactId>cmake-maven-plugin</artifactId>
```
Comment (SammyVimes) [inline modules/numa-allocator/pom.xml:110]: This plugin will be downloading cmake every time and I believe we have cmake installed on teamcity. 
Consider adding 
```
<configuration>
    <downloadBinaries>false</downloadBinaries>
</configuration>
```

File: modules/numa-allocator/src/main/cpp/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,44 @@
+#
+# Copyright 2019 GridGain Systems, Inc. and Contributors.
```
Comment (SammyVimes) [inline modules/numa-allocator/src/main/cpp/CMakeLists.txt:2]:
```suggestion
# Copyright 2022 GridGain Systems, Inc. and Contributors.
```

File: modules/numa-allocator/src/main/cpp/include/numa/numa_alloc.h [hunk 6]
```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright 2019 GridGain Systems, Inc. and Contributors.
```
Comment (SammyVimes) [inline modules/numa-allocator/src/main/cpp/include/numa/numa_alloc.h:2]:
```suggestion
 * Copyright 2022 GridGain Systems, Inc. and Contributors.
```

File: modules/numa-allocator/src/main/cpp/include/org_apache_ignite_internal_mem_NumaAllocUtil.h [hunk 7]
```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright 2019 GridGain Systems, Inc. and Contributors.
```
Comment (SammyVimes) [inline modules/numa-allocator/src/main/cpp/include/org_apache_ignite_internal_mem_NumaAllocUtil.h:2]:
```suggestion
 * Copyright 2022 GridGain Systems, Inc. and Contributors.
```

File: modules/numa-allocator/src/main/cpp/src/numa/numa_alloc.cpp [hunk 8]
```diff
@@ -0,0 +1,213 @@
+/*
+ * Copyright 2019 GridGain Systems, Inc. and Contributors.
```
Comment (SammyVimes) [inline modules/numa-allocator/src/main/cpp/src/numa/numa_alloc.cpp:2]:
```suggestion
 * Copyright 2022 GridGain Systems, Inc. and Contributors.
```

File: modules/numa-allocator/src/main/cpp/src/org_apache_ignite_internal_mem_NumaAllocUtil.cpp [hunk 9]
```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright 2019 GridGain Systems, Inc. and Contributors.
```
Comment (SammyVimes) [inline modules/numa-allocator/src/main/cpp/src/org_apache_ignite_internal_mem_NumaAllocUtil.cpp:2]:
```suggestion
 * Copyright 2022 GridGain Systems, Inc. and Contributors.
```

File: modules/numa-allocator/src/main/java/org/apache/ignite/internal/mem/NumaAllocUtil.java [hunk 10]
```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright 2019 GridGain Systems, Inc. and Contributors.
```
Comment (SammyVimes) [inline modules/numa-allocator/src/main/java/org/apache/ignite/internal/mem/NumaAllocUtil.java:2]:
```suggestion
 * Copyright 2022 GridGain Systems, Inc. and Contributors.
```

File: modules/numa-allocator/src/main/java/org/apache/ignite/internal/mem/NumaAllocUtil.java [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+    }
+
+    /** */
+    private static String getOSArch() {
+        return System.getProperty("os.arch", "");
+    }
+
+    /** */
+    private static String getOSName() {
+        if (System.getProperty("os.name", "").contains("Linux"))
```
Comment (SammyVimes) [inline modules/numa-allocator/src/main/java/org/apache/ignite/internal/mem/NumaAllocUtil.java:51]: Maybe just check the OS to be linux in the static block then? And in `getLibName()` method just use "linux"

File: modules/numa-allocator/src/main/java/org/apache/ignite/mem/InterleavedNumaAllocationStrategy.java [hunk 12]
```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright 2019 GridGain Systems, Inc. and Contributors.
```
Comment (SammyVimes) [inline modules/numa-allocator/src/main/java/org/apache/ignite/mem/InterleavedNumaAllocationStrategy.java:2]:
```suggestion
 * Copyright 2022 GridGain Systems, Inc. and Contributors.
```

File: modules/numa-allocator/src/main/java/org/apache/ignite/mem/LocalNumaAllocationStrategy.java [hunk 13]
```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright 2019 GridGain Systems, Inc. and Contributors.
```
Comment (SammyVimes) [inline modules/numa-allocator/src/main/java/org/apache/ignite/mem/LocalNumaAllocationStrategy.java:2]:
```suggestion
 * Copyright 2022 GridGain Systems, Inc. and Contributors.
```

File: modules/numa-allocator/src/main/java/org/apache/ignite/mem/LocalNumaAllocationStrategy.java [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+
+/**
+ * Local NUMA allocation strategy.
+ * <p>
+ * Memory will be allocated using {@code void* numa_alloc_onnode(size_t)} of {@code lubnuma}.
+ */
+public class LocalNumaAllocationStrategy implements NumaAllocationStrategy, Serializable {
+    /**
+     *
+     */
```
Comment (SammyVimes) [inline modules/numa-allocator/src/main/java/org/apache/ignite/mem/LocalNumaAllocationStrategy.java:31]:
```suggestion
    /** */
```

File: modules/numa-allocator/src/main/java/org/apache/ignite/mem/NumaAllocationStrategy.java [hunk 15]
```diff
@@ -0,0 +1,29 @@
+/*
+ * Copyright 2019 GridGain Systems, Inc. and Contributors.
```
Comment (SammyVimes) [inline modules/numa-allocator/src/main/java/org/apache/ignite/mem/NumaAllocationStrategy.java:2]:
```suggestion
 * Copyright 2022 GridGain Systems, Inc. and Contributors.
```

File: modules/numa-allocator/src/main/java/org/apache/ignite/mem/NumaAllocator.java [hunk 16]
```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright 2019 GridGain Systems, Inc. and Contributors.
```
Comment (SammyVimes) [inline modules/numa-allocator/src/main/java/org/apache/ignite/mem/NumaAllocator.java:2]:
```suggestion
 * Copyright 2022 GridGain Systems, Inc. and Contributors.
```

File: modules/numa-allocator/src/main/java/org/apache/ignite/mem/SimpleNumaAllocationStrategy.java [hunk 17]
```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2019 GridGain Systems, Inc. and Contributors.
```
Comment (SammyVimes) [inline modules/numa-allocator/src/main/java/org/apache/ignite/mem/SimpleNumaAllocationStrategy.java:2]:
```suggestion
 * Copyright 2022 GridGain Systems, Inc. and Contributors.
```

File: modules/numa-allocator/src/test/java/org/apache/ignite/internal/mem/NumaAllocatorBasicTest.java [hunk 18]
```diff
@@ -0,0 +1,176 @@
+/*
+ * Copyright 2019 GridGain Systems, Inc. and Contributors.
```
Comment (SammyVimes) [inline modules/numa-allocator/src/test/java/org/apache/ignite/internal/mem/NumaAllocatorBasicTest.java:2]:
```suggestion
 * Copyright 2022 GridGain Systems, Inc. and Contributors.
```

File: modules/numa-allocator/src/test/java/org/apache/ignite/internal/mem/NumaAllocatorUnitTest.java [hunk 19]
```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright 2019 GridGain Systems, Inc. and Contributors.
```
Comment (SammyVimes) [inline modules/numa-allocator/src/test/java/org/apache/ignite/internal/mem/NumaAllocatorUnitTest.java:2]:
```suggestion
 * Copyright 2022 GridGain Systems, Inc. and Contributors.
```

File: modules/numa-allocator/src/test/java/org/apache/ignite/testsuites/NumaAllocatorTestSuite.java [hunk 20]
```diff
@@ -0,0 +1,31 @@
+/*
+ * Copyright 2019 GridGain Systems, Inc. and Contributors.
```
Comment (SammyVimes) [inline modules/numa-allocator/src/test/java/org/apache/ignite/testsuites/NumaAllocatorTestSuite.java:2]:
```suggestion
 * Copyright 2022 GridGain Systems, Inc. and Contributors.
```

**Review by SammyVimes — CHANGES_REQUESTED** [review]

File: modules/numa-allocator/src/main/java/org/apache/ignite/internal/mem/NumaAllocUtil.java (hunk 11 above)
Comment (ademakov) [reply modules/numa-allocator/src/main/java/org/apache/ignite/internal/mem/NumaAllocUtil.java:51]: I think we should keep the code identical to the code in Ignite unless we can get some real improvement. I am in doubt if your suggested change improves things. This method will be very rarely used. Why bother optimizing it?

Comment (ademakov) [reply modules/numa-allocator/src/main/java/org/apache/ignite/internal/mem/NumaAllocUtil.java:51]: Done.

File: modules/numa-allocator/pom.xml (hunk 4 above)
Comment (ademakov) [reply modules/numa-allocator/pom.xml:110]: Done.

**Review by SammyVimes — APPROVED** [review]
LGTM

**Pull request merged**